python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py

```diff
@@ -333,75 +422,141 @@ def terminate(self, run_id):
         self.ecs.stop_task(task=tags.arn, cluster=tags.cluster)
         return True
 
-    def _task_definition(self, family, metadata, image, container_context):
+    def _task_config(self, run, image, container_context) -> DagsterEcsTaskConfig:
         """
-        Return the launcher's task definition if it's configured.
-
-        Otherwise, a new task definition revision is registered for every run.
-        First, the process that calls this method finds its own task
-        definition. Next, it creates a new task definition based on its own
-        but it overrides the image with the pipeline origin's image.
+        Return a DagsterEcsTaskConfig to use to launch the ECS task, registering a new task
+        definition if needed.
         """
+        environment = self._environment(container_context)
+        secrets = self._secrets(container_context)
+
         if self.task_definition:
+            # Use the passed in task definition as a basis for the launched task (using the current
+            # ECS task to determine network configuration and security groups)
+            current_task_metadata = get_current_ecs_task_metadata()
+            current_task = get_current_ecs_task(
+                self.ecs, current_task_metadata.task_arn, current_task_metadata.cluster
+            )
             task_definition = self.ecs.describe_task_definition(taskDefinition=self.task_definition)
-            return task_definition["taskDefinition"]
 
-        environment = [
-            {"name": key, "value": value}
-            for key, value in container_context.get_environment_dict().items()
-        ]
+            return get_task_config_from_current_task(
+                self.ec2,
+                current_task_metadata.cluster,
+                current_task,
+                task_definition["taskDefinition"]["family"],
+            )
 
-        secrets = container_context.get_secrets_dict(self.secrets_manager)
-        secrets_definition = (
-            {"secrets": [{"name": key, "valueFrom": value} for key, value in secrets.items()]}
-            if secrets
-            else {}
+        family = sanitize_family(
+            run.external_pipeline_origin.external_repository_origin.repository_location_origin.location_name  # type: ignore
         )
 
-        task_definition = {}
-        with suppress(ClientError):
-            task_definition = self.ecs.describe_task_definition(taskDefinition=family)[
+        if self.use_current_task_definition:
```

Comment:
What if we instead let people set an optional "TaskDefinition" dict in config - and we merged that dict with our existing defaults? That might help squash "impossible" states where you set `use_current_task_definition` to `False` but also don't provide everything needed to define a new task definition.

I'm also just having difficulty following what even happens differently if you set this to `True` - what do we still pull from the parent's task definition and not from the launcher config that couldn't be represented as a "default" value?

Reply:
i'm excited to offer something like a task_definition_kwargs field or something that lets you specify arbitrary task definition config so that we don't have to re-implement the whole spec as Dagster config, but I don't think it would be sufficient here. The pieces config that we're offering are a mixture of things on the dagster container (log configuration / sidecars), the task definition (execution_role_arn, task_role_arn), and the task (cluster / security groups / subnets).

Generally I'm trying to follow the general playbook here (https://elementl.quip.com/9K0bAWhc6t9n/Agent-Configuration-Options) - where we have a relatively simple and blessed way to set the most common pieces of config we think people are likely to want(secrets etc.) and then eventually also an escape hatch where we expose the full raw config for the power users. I think there's value in not needing to send people to the boto spec if they want to do something like 'set env vars' or 'point at an existing cluster' or 'change the IAM role that's used'. But which things are considered 'the most common pieces of config' is very subjective.